atsign-foundation/at_server
# feat: reset handling fix 1.2a

**What I did**
- refactor: Add AtCacheManager so that we can handle all caching operations in one place
- refactor: Move cache-related operations from LookupVerbHandler and ProxyLookupVerbHandler into AtCacheManager
- test: Added unit tests covering full behaviour of LookupVerbHandler and ProxyLookupVerbHandler including caching
- feat: Handle resets of other atSigns by detecting changes to their public encryption keys
- test: Added unit tests covering behaviour when public encryption keys changes detected
- test: Added unit tests covering behaviour of the CacheRefreshJob

**How I did it**
The material code changes are in
- AtCacheManager
- OutboundClient
- AtCacheRefreshJob

The most interesting code is in the 34 new unit tests spread across
- lookup_verb_test.dart
- proxy_lookup_verb_test.dart
- cache_refresh_job_test.dart
- reset_handling_test.dart
- and a huge amount of shared and reusable mock setup boilerplate in test_utils.dart 

All of the other code changes are lint removal and other zero-impact refactoring
 
**How I verified it**
- [x] Automated tests pass in at_server
- [x] Manual testing using at_talk
  - [x] Run alice and bob servers using server version 3.0.27; start chatting; reset bob server; observe that nothing gets through from alice to bob thereafter
  - [x] Run the servers using server version from this branch; observe that alice client creates a new shared key, and stuff starts to get through from alice to bob again
- [x] Run the at_client_sdk end-to-end test pack against servers running from this branch
  - This was complicated! Am including the steps required here.. Thanks to @purnimavenkatasubbu and @sitaram-kalluri for your help
    - Run two new at_servers alice and bob (at_server script)
    - Run onboard against both to generate atKeys (at_onboard script)
    - Copy the keys to a tmp local directory
      - `cd ~/dev/atsign/repos/at_client_sdk/tests/at_end2end_test`
      - mkdir tmp_keys
      - cp ~/.atsign/keys/@alice_key.atKeys tmp_keys
      - cp ~/.atsign/keys/@bob_key.atKeys tmp_keys
    - run at_generate_demo_data.dart against that directory to create at_credentials.dart
      - `dart ~/dev/atsign/repos/at_tools/packages/at_dump_atKeys/bin/generate_at_demo_data.dart -d tmp_keys  -f at_credentials.dart`
    - move the generated at_credentials.dart to the correct place in the end2end tests pack and modify it to correct format
      - `mv at_credentials.dart lib/src`
    - Modify the other config in the end2end tests pack
      - tests/at_end2end_test/config/config12.yaml
      - at_end2end_test/lib/config/config_util.dart
    - Run the client end2end tests
      - dart test --concurrency=1
